src/main/java/com/fdmgroup/forex/services/FxRateService.java

```diff
@@ -1,38 +1,135 @@
+    }
+
+    private String fetchFxRatesFromAPI() throws IOException {
+        URL obj = new URL(fxApiUsdUrl);
+        HttpURLConnection con = (HttpURLConnection) obj.openConnection();
+        con.setRequestMethod("GET");
+        int responseCode = con.getResponseCode();
+
+        if (responseCode == HttpURLConnection.HTTP_OK) {
+            BufferedReader input = new BufferedReader(new InputStreamReader(con.getInputStream()));
+            StringBuilder response = new StringBuilder();
+            String inputLine;
+            while ((inputLine = input.readLine()) != null) {
+                response.append(inputLine);
+            }
+            input.close();
+            return response.toString();
+        } else {
+            throw new IOException("Failed to fetch exchange rates from API");
+        }
+    }
+
+    private Map<String,Double> parseFxRatesResponse(String response) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        Map<String, Object> data = mapper.readValue(response, new TypeReference<Map<String, Object>>() {});
+
+        Object ratesObject = data.get("rates");
+        if (!(ratesObject instanceof Map<?, ?>)) {
+            throw new IllegalArgumentException("Expected 'rates' to be a Map, but got: " + ratesObject.getClass());
+        }
+    
+        Map<?, ?> rawRates = (Map<?, ?>) ratesObject;
+        Map<String, Double> rates = new HashMap<>();
+    
+        for (Map.Entry<?, ?> entry : rawRates.entrySet()) {
+            if (!(entry.getKey() instanceof String)) {
+                throw new IllegalArgumentException("Expected rate key to be a String, but got: " + entry.getKey().getClass());
+            }
+            String key = (String) entry.getKey();
+            Object value = entry.getValue();
+    
+            if (value instanceof Integer intValue) {
+                rates.put(key, intValue.doubleValue());
+            } else if (value instanceof Double doubleValue) {
+                rates.put(key, doubleValue);
+            } else if (value instanceof Number numberValue) {
+                rates.put(key, numberValue.doubleValue());
+            } else {
+                throw new IllegalArgumentException("Unexpected value type in rates map: " + value.getClass());
+            }
+        }
+        return rates;
+    }
+
+    private List<FxRate> updateFxRatesFromUSDRates(Map<String,Double> rates) {
+        List<FxRate> fxRates = new ArrayList<>();
+        for (String code : rates.keySet()) {
+            double rate = Double.valueOf(rates.get(code));
+            try {
+                FxRate fxRate = createOrUpdateFxRate(code, rate);
+                fxRates.add(fxRate);
+            } catch (RecordNotFoundException e) {
+
+            }
+        }
+        return fxRates;
+    }
+
+    private FxRate createOrUpdateFxRate(String code, double rate) {
+        try {
+            FxRate fxRate = findFxRateByCurrencyId(code);
+            fxRate.setRateToUSD(1.0 / rate);
```

Comment:
Do we need to consider the possibility of rate equals 0 or less than 0? And handle it gracefully?

Reply:
0 should not be possible - I think we can trust the API to provide only valid exchange rates